src/ray/object_manager/connection_pool.h

```diff
@@ -0,0 +1,148 @@
+#ifndef RAY_CONNECTION_POOL_H
+#define RAY_CONNECTION_POOL_H
+
+#include <algorithm>
+#include <cstdint>
+#include <deque>
+#include <map>
+#include <memory>
+#include <thread>
+
+#include <boost/asio.hpp>
+#include <boost/asio/error.hpp>
+#include <boost/bind.hpp>
+
+#include "ray/id.h"
+#include "ray/status.h"
+
+#include "format/object_manager_generated.h"
+#include "object_directory.h"
+#include "object_manager_client_connection.h"
+
+namespace asio = boost::asio;
+
+namespace ray {
+
+class ConnectionPool {
```

Comment:
This definition for the `ConnectionPool` is doing a lot and we should think more carefully about what its exact purpose is going to be, especially since it will likely be shared by multiple threads. I think a cleaner design is to limit the `ConnectionPool`'s purpose to only storing *inactive outgoing connections*.

Reply:
This is a good point and something I thought through when deciding the role of the `ConnectionPool`. I opted for encapsulation, which broadens the functionality of this class beyond just simply maintaining a pool of outgoing connections. The current design simplifies the `ObjectManager` implementation.